### PR TITLE
(PUP-4030) Support no_proxy env variable

### PR DESCRIPTION
**Pull request by ChrisPortman**
Implement support for the no_proxy variable in various places where http_proxy was taken into consideration such as accessing the forge.

### CODE REVIEW

File: lib/puppet/util/http_proxy.rb [hunk 1]
```diff
@@ -11,6 +20,34 @@ def self.http_proxy_env
...
+        return false
+      end
+    end
+
+    no_proxy_env.split(/\s*,\s*/).each do |d|
+      host, port = d.split(':')
+      host = Regexp.escape(host).gsub('\*', '.*')
+
+      regex  = port ? %r(^#{host}:#{port}$) : %r(^#{host}$)
+      string = port ? "#{dest.host}:#{dest.port}" : "#{dest.host}"
```
Comment (ffrank) [inline lib/puppet/util/http_proxy.rb]: Nitpick: The variable names `regex` and `string` make it hard to decipher the intent of these lines and the next.

File: lib/puppet/network/http/factory.rb [hunk 2]
```diff
@@ -23,10 +23,13 @@ def create_connection(site)
...
-    if Puppet[:http_proxy_host] == "none"
-      args << nil << nil
-    else
-      args << Puppet[:http_proxy_host] << Puppet[:http_proxy_port]
+
+    unless Puppet::Util::HttpProxy.no_proxy?(site)
+      if Puppet[:http_proxy_host] == "none"
+        args << nil << nil
+      else
+        args << Puppet[:http_proxy_host] << Puppet[:http_proxy_port]
```
Comment (joshcooper) [inline lib/puppet/network/http/factory.rb]: It appears puppet's http connection factory doesn't handle `http_proxy_user` and `http_proxy_password`. Could you file a new ticket for that? And even submit a separate PR for that?

**Comment by joshcooper** [comment]
@ChrisPortman thank you for your contribution! During PR triage we had a few comments:

* Could you add unit tests for the other parts of code that were changed, `repository`, `factory`, `pkgdmg`
* In general we try to break up PRs like this into logical commits, e.g. update http_proxy to have a `no_proxy?` method and add tests, then another commit to modify the callers to take that into account. I wouldn't say it's a requirement for this PR, but it makes it easier to reason about the changes.

**Comment by branan** [comment]
@ChrisPortman Hi, it's been a couple of weeks since we commented on this.

If we don't hear from you in 2-3 more weeks we'll be closing this PR for inactivity.

**Comment by branan** [comment]
@ChrisPortman Will you have a chance to address Josh's comments here?

**Comment by ChrisPortman** [comment]
Hi All,

Apologies, I've been on leave the last couple of weeks.  I will be able to address the feedback, hopefully this week.

Thanks,
Chris

File: lib/puppet/util/http_proxy.rb (hunk 1 above)
Comment (ChrisPortman) [inline lib/puppet/util/http_proxy.rb]: Hi @ffrank 

I've updated the var names to be something a little more descriptive.  I've also added some commentary regarding the intention here.

Thanks.
Chris

**Comment by ChrisPortman** [comment]
Hi @joshcooper 

Thanks for your feedback.  I did look for opportunities to add spec tests to the other areas when I made the changes, but I couldn't find a way to add any meaningful tests as I didn't change the API in the other cases (i.e the inputs and outputs are the same).

That said, my experience is limited.  So I'd appreciate any pointers if you have any.

Also, I see your point regarding a commit per change.  I'll definitely keep it in mind in future.

Thanks,
Chris

File: lib/puppet/util/http_proxy.rb [hunk 3]
```diff
@@ -11,6 +20,36 @@ def self.http_proxy_env
...
+    end
+
+    no_proxy_env.split(/\s*,\s*/).each do |d|
+      host, port = d.split(':')
+      host = Regexp.escape(host).gsub('\*', '.*')
+
+      #If this no_proxy entry specifies a port, we want to match it against
+      #the destination port.  Otherwise just match hosts.
+      no_proxy_regex  = port ? %r(^#{host}:#{port}$) : %r(^#{host}$)
+      dest_string     = port ? "#{dest.host}:#{dest.port}" : "#{dest.host}"
```
Comment (ffrank) [inline lib/puppet/util/http_proxy.rb]: Thanks, much more readable to me! :+1:

Now that I grokked it, I come to believe that a

    if port
      no_proxy_regex = ...
      dest_string = ...
    else
      ...
    end

would be even clearer. Thoughts?

Comment (ChrisPortman) [inline lib/puppet/util/http_proxy.rb]: Hi @ffrank 

Agreed, I updates the code to match.

File: lib/puppet/provider/package/pkgdmg.rb [hunk 4]
```diff
@@ -68,8 +68,10 @@ def self.installpkg(source, name, orig_source)
   end
 
   def self.installpkgdmg(source, name)
-    http_proxy_host = Puppet::Util::HttpProxy.http_proxy_host
-    http_proxy_port = Puppet::Util::HttpProxy.http_proxy_port
+    unless Puppet::Util::HttpProxy.no_proxy?(source)
```
Comment (branan) [inline lib/puppet/provider/package/pkgdmg.rb]: There are spec tests for setting the proxy here. Could you expand those to add `no_proxy` coverage?

File: lib/puppet/network/http/factory.rb [hunk 5]
```diff
@@ -23,10 +23,13 @@ def create_connection(site)
     Puppet.debug("Creating new connection for #{site}")
 
     args = [site.host, site.port]
-    if Puppet[:http_proxy_host] == "none"
-      args << nil << nil
-    else
-      args << Puppet[:http_proxy_host] << Puppet[:http_proxy_port]
+
+    unless Puppet::Util::HttpProxy.no_proxy?(site)
```
Comment (branan) [inline lib/puppet/network/http/factory.rb]: Same as for pkgdmg - there are existing tests for proxy management here, it would be cool to see them expanded to cover the no proxy case.

File: lib/puppet/provider/package/pkgdmg.rb (hunk 4 above)
Comment (ChrisPortman) [inline lib/puppet/provider/package/pkgdmg.rb]: Hi @branan ,

Ahh I see the tests you're referring to,  Ive added an additional test to account for the no_proxy variable.

File: lib/puppet/network/http/factory.rb (hunk 5 above)
Comment (ChrisPortman) [inline lib/puppet/network/http/factory.rb]: I've also augmented the tests here to account for no_proxy.

**Comment by ChrisPortman** [comment]
I've just rebased it against the current master as well.

**Comment by peterhuene** [comment]
@ChrisPortman Thanks for the new commit.  Could you amend the commit message to include the ticket number?

**Comment by melissa** [comment]
@ChrisPortman for more information about the commit message format, check out the documentation at https://github.com/puppetlabs/puppet/blob/master/CONTRIBUTING.md#making-changes

**Comment by ChrisPortman** [comment]
OK, i've update the commit message (sorry for some reason I thought the PR message was what was included as the merge commit).

Also, it came to my attention over the weekend, that depending on where you look, domain level no_proxy entries are either '*.example.com' or '.example.com', so I've updated the code to support either notation.

File: lib/puppet/network/http/factory.rb (hunk 2 above)
Comment (joshcooper) [inline lib/puppet/network/http/factory.rb]: I filed https://tickets.puppetlabs.com/browse/PUP-4470 for this

**Comment by fgimian** [comment]
This pull request will really make a huge difference for us, thanks!! Really hoping it is accepted :smile:

**Comment by peterhuene** [comment]
I'm :+1:

**Comment by peterhuene** [comment]
@ChrisPortman Thanks again for the contribution!  This has been merged to master and will appear in the next release of Puppet.

**Comment by fgimian** [comment]
Thank you so much everyone!!! :smile:
